# MultiAlign

MultiAlign aligns multiple LC-MS datasets to one another after which LC-MS 
features can be matched to a database of peptides (typically an 
AMT tag database).  The program reads deisotoped features from either
[DeconTools](https://github.com/PNNL-Comp-Mass-Spec/DeconTools/releases) or ICR-2LS.
The LC-MS/MS Accurate Mass and Time Tag database 
is a SQLite file, typically generated by [MTDB Creator](https://github.com/PNNL-Comp-Mass-Spec/MTDB-Creator/releases)

## Features

The MultiAlign application provides advanced visualization and manipulation 
capabilities for LC-MS datasets acquired on high resolution mass 
spectrometers. Functionalities include overlaid 2D plots, alignment plots, 
normalizations, and basic statistical comparisons. 

MultiAlign uses the [LCMSWARP algorithm](https://github.com/PNNL-Comp-Mass-Spec/MultiAlign/tree/master/src/Library/MultiAlignCore/Algorithms/Alignment/LcmsWarp)
to align LC-MS datasets to a master list (mass tag database or a single LC-MS dataset) 
then consolidates features into a consensus map. 
These features can be matched to an AMT tag database to identify the LC-MS 
features. Data processing is similar to that performed with the [VIPER](https://github.com/PNNL-Comp-Mass-Spec/VIPER)
application, but MultiAlign has the advantage of finding consensus features 
followed by matching to an AMT tag DB. The processing performed by MultiAlign 
results in fewer "missing values" vs. the results obtained when analyzing 
separate datasets with VIPER.

## Contacts

Written by Brian LaMarche, Bryson Gibbons, and Chris Wilkins for the Department of Energy (PNNL, Richland, WA) \
E-mail: proteomics@pnnl.gov \
Website: https://panomics.pnl.gov/ or https://omics.pnl.gov

## License

MultiAlign is licensed under the Apache License, Version 2.0; 
you may not use this file except in compliance with the License.  You may obtain 
a copy of the License at https://opensource.org/licenses/Apache-2.0
